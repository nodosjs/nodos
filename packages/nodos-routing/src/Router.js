import _ from 'lodash';
import urlJoin from 'url-join';
import Route from './Route';
import validate from './validator';

const detectRouteType = (currentName) => {
  const names = ['resources', 'resource'];
  return names.find(name => name === currentName);
};

const routesDefaultOnly = ['index', 'new', 'create', 'show', 'edit', 'update', 'destroy'];
const routesDefaultExcept = [];

const buildActionNames = (routeItem) => {
  const onlyNames = _.get(routeItem, 'only', routesDefaultOnly);
  const exceptNames = _.get(routeItem, 'except', routesDefaultExcept);
  return _.difference(onlyNames, exceptNames);
};

const normalizeRouteItem = (valueOrValues) => {
  const routeItem = _.isObject(valueOrValues) ? valueOrValues : { name: valueOrValues };

  return {
    ...routeItem,
    actionNames: buildActionNames(routeItem),
  };
};

const selectRequestedActions = (routeItem, actions) => {
  const { actionNames } = routeItem;
  return actions.filter(action => actionNames.includes(action.name));
};

const types = {
  resources: (routeItem, rec, { path, middlewares, parent }) => {
    const getParentPath = (parent) => {
      const { path, resourceName } = parent;
      return urlJoin(path, resourceName, `:${resourceName.slice(0, -1)}_id`);
    };

    const buildUrl = u => urlJoin(path, routeItem.name, u);
    const actions = [
      { name: 'index', method: 'get', url: buildUrl('') },
      { name: 'new', method: 'get', url: buildUrl('new') },
      { name: 'create', method: 'post', url: buildUrl('') },
      { name: 'show', method: 'get', url: buildUrl(':id') },
      { name: 'edit', method: 'get', url: buildUrl(':id/edit') },
      { name: 'update', method: 'patch', url: buildUrl(':id') },
      { name: 'update', method: 'put', url: buildUrl(':id') },
      { name: 'destroy', method: 'delete', url: buildUrl(':id') },
    ];

    const route = new Route({
      middlewares,
      path,
      parent,
      resourceName: routeItem.name,
      actions: selectRequestedActions(routeItem, actions),
    });

    const nestedRoutes = routeItem.routes
      ? buildRoutes(routeItem.routes, { path: getParentPath(route), middlewares, parent: route })
      : [];

    return [route, ...nestedRoutes];
  },
  resource: (routeItem, rec, { path, middlewares, parent }) => {
    const buildUrl = u => urlJoin(path, routeItem.name, u);
    const actions = [
      { name: 'new', method: 'get', url: buildUrl('new') },
      { name: 'create', method: 'post', url: buildUrl('') },
      { name: 'show', method: 'get', url: buildUrl('') },
      { name: 'edit', method: 'get', url: buildUrl('edit') },
      { name: 'update', method: 'patch', url: buildUrl('') },
      { name: 'update', method: 'put', url: buildUrl('') },
      { name: 'destroy', method: 'delete', url: buildUrl('') },
    ];

    const route = new Route({
      middlewares,
      path,
      parent,
      resourceName: routeItem.name,
      actions: selectRequestedActions(routeItem, actions),
    });

    const nestedRoutes = routeItem.routes
      ? buildRoutes(routeItem.routes, { path: buildUrl(''), middlewares, parent: route })
      : [];

    return [route, ...nestedRoutes];
  },
};

export const buildRoutes = (routes, options) => routes.map((item) => {
  const typeName = detectRouteType(Object.keys(item)[0]);
  const routeItem = normalizeRouteItem(item[typeName]);
  return types[typeName](routeItem, buildRoutes, options);
});

const buildScope = ({ routes, path, pipeline }, pipelines) => {
  const result = buildRoutes(routes, { path, middlewares: pipelines[pipeline] });
  return result;
};

export default class Router {
  constructor(routeMap) {
    validate(routeMap);

    this.routeMap = routeMap;
    this.routes = routeMap.scopes.map(scope => buildScope(scope, routeMap.pipelines))
      |> _.flattenDeep;
  }

  // recognize(request) {
  // }
}
