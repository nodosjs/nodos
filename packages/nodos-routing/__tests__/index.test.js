import { promises as fs } from 'fs';
import _ from 'lodash';
import yml from 'js-yaml';
import Router from '../src';

test('nodos-routing', async () => {
  const routesData = await fs.readFile(`${__dirname}/__fixtures__/routes.yml`);
  const routesMap = yml.safeLoad(routesData);
  const { routes } = new Router(routesMap);

  const expectedRoutes = [
    { name: 'index', url: '/api/users', method: 'get' },
    { name: 'new', url: '/api/users/new', method: 'get' },
    { name: 'create', url: '/api/users', method: 'post' },
    { name: 'show', url: '/api/users/:id', method: 'get' },
    { name: 'show', url: '/users/:id', method: 'get' },
    { name: 'show', url: '/session', method: 'get' },
    { name: 'update', url: '/session', method: 'patch' },
    { name: 'show', url: '/session/tokens/:id', method: 'get' },
    { name: 'index', url: '/articles', method: 'get' },
    { name: 'update', url: '/articles/:id', method: 'patch' },
    { name: 'index', url: '/articles/:article_id/comments', method: 'get' },
    { name: 'show', url: '/articles/:article_id/comments/:id', method: 'get' },
  ];

  const actualRoutes = routes.flatMap(({ actions }) => actions)
    .filter(r => expectedRoutes.find(e => _.isEqual(e, r)));

  expect(_.uniqWith(actualRoutes, _.isEqual)).toEqual(expectedRoutes);

  // Only
  const userRootRoutes = _.find(routes, { resourceName: 'users', path: '/' });

  expect(userRootRoutes.actions).toHaveLength(1);
  expect(userRootRoutes.actions[0]).toMatchObject({
    method: 'get',
    name: 'show',
    url: '/users/:id',
  });

  // Except
  const sessionRootRoutes = _.find(routes, { resourceName: 'session' });
  const sessionNewRoute = _.find(sessionRootRoutes, { name: 'new' });
  expect(sessionNewRoute).toBeUndefined();
});

test('nodos-routing throws an error if schema is invalid', async () => {
  const routesData = await fs.readFile(`${__dirname}/__fixtures__/routesWithInvalidSchema.yml`);
  const routesMap = yml.safeLoad(routesData);

  expect(() => new Router(routesMap)).toThrowErrorMatchingSnapshot();
});
