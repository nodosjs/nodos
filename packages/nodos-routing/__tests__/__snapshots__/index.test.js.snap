// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nodos Routing nodos-routing throws an error if schema is invalid 1`] = `"invalidRoutesMap is not defined"`;

exports[`nodos-routing throws an error if schema is invalid 1`] = `"Routes schema is invalid"`;
